src/librustc/ty/layout.rs

```diff
@@ -1111,47 +1079,50 @@ impl<'tcx> LayoutCx<'tcx, TyCtxt<'tcx>> {
                     // after this point – we’ll just truncate the value we load in codegen.
                 }
 
-                // Check to see if we should use a different type for the
-                // discriminant. We can safely use a type with the same size
-                // as the alignment of the first field of each variant.
+                if min_ity.size() > gap {
+                    gap += (min_ity.size() - gap).align_to(align.abi);
+                }
+
                 // We increase the size of the discriminant to avoid LLVM copying
                 // padding when it doesn't need to. This normally causes unaligned
                 // load/stores and excessive memcpy/memset operations. By using a
                 // bigger integer size, LLVM can be sure about its contents and
                 // won't be so conservative.
-
-                // Use the initial field alignment
-                let mut ity = if def.repr.c() || def.repr.int.is_some() {
+                let ity = if def.repr.inhibit_enum_layout_opt() {
                     min_ity
                 } else {
-                    Integer::for_align(dl, start_align).unwrap_or(min_ity)
+                    Integer::approximate_size(gap).unwrap()
```

Comment:
If you pick something large enough to increase the alignment, you'll easily waste space in types containing this one.
If you think this will never happen, that's still not obvious from the code, and you'd have to either compute both and assert they're the same, and we can land that, crater it, etc. *or* prove it formally somehow.

Reply:
Let `gap0` be the original value of `gap`. We either have `gap = gap0 < align.abi`, or `gap = gap0 + (min_ity.size() - gap0).align_to(align.abi) < gap0 + min_ity.size() - gap0 + align.abi = min_ity.size() + align.abi`.

`ity.size() <= gap`. In both cases `ity.size() <= gap < min_ity.size() + align.abi`, therefore either `ity = min_ity` or `ity.size() < align.abi`.